### ISSUE TITLE
Is Sampler forward() working properly now?

### ISSUE THREAD

**Issue by yhlskt23** [vllm-project/vllm#1179]
Hello.
I'm using the vLLM very well.
But in a recent commit, I find some code that may have bugs.

I suspect there is a bug in the way of applying sampling_params in the vllm/model_executor/layers/sampler.py.

Watching a _prune_hidden_states function , It seems that logits is sorted by the SamplingType.
However, when getting the sampling parameters, those ​​are just ordered by the input_metadata.seq_groups. (e.g. _get_penalties)
I can't find a way to match these two in a step of applying parameters. (e.g. _apply_penalties)
It may be happened that sampling parameters are applied to wrong logits.

Please let me know if I misunderstood anything.
